bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/EventFilterHandler.java

```diff
@@ -0,0 +1,362 @@
+                Channel currentChannel = builder.build();
+                @Nullable
+                Channel existingChannel = getThing().getChannel(currentChannel.getUID());
+                if (existingChannel == null) {
+                    thingBuilder.withChannel(currentChannel);
+                }
+            }
+        });
+
+        updateThing(thingBuilder.build());
+    }
+
+    /**
+     * Updates all states and channels.
+     */
+    private void updateStates() {
+        final Bridge iCalendarBridge = getBridge();
+        if (iCalendarBridge == null) {
+            logger.debug("Bridge not instantiated!");
+            return;
+        }
+        final ICalendarHandler iCalendarHandler = (ICalendarHandler) iCalendarBridge.getHandler();
+        if (iCalendarHandler == null) {
+            logger.debug("ICalendarHandler not instantiated!");
+            return;
+        }
+        final EventFilterConfiguration config = configuration;
+        if (config == null) {
+            logger.debug("Configuration not instantiated!");
+            return;
+        }
+        @Nullable
+        final AbstractPresentableCalendar cal = iCalendarHandler.getRuntimeCalendar();
+        if (cal != null) {
+            updateStatus(ThingStatus.ONLINE);
+
+            String textFilterValue = config.textEventValue;
+            EventTextFilter filter = null;
+            if (textFilterValue != null) {
+                try {
+                    if (config.textEventField == null || config.textValueType == null) {
+                        throw new IllegalArgumentException("Config is broken.");
+                    }
+                    EventTextFilter.Field textFilterField = EventTextFilter.Field.valueOf(config.textEventField);
+                    EventTextFilter.Type textFilterType = EventTextFilter.Type.valueOf(config.textValueType);
+
+                    filter = new EventTextFilter(textFilterField, textFilterValue, textFilterType);
+                } catch (IllegalArgumentException e) {
+                    logger.warn("Text-filter settings are not set properly.");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+                    return;
+                }
+            }
+
+            Instant reference = Instant.now();
+            TimeMultiplicator multiplicator;
+
+            try {
+                if (config.datetimeUnit == null) {
+                    throw new IllegalArgumentException("Config is broken.");
+                }
+                multiplicator = TimeMultiplicator.valueOf(config.datetimeUnit);
+            } catch (IllegalArgumentException e) {
+                logger.warn("Time-filter settings are not set properly.");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+                return;
+            }
+
+            if (config.datetimeRound) {
+                ZonedDateTime refDT = reference.atZone(ZoneId.systemDefault());
```

Comment:
You could inject the `TimeZoneProvider` to retrieve the timezone configured within openHAB. This may differ from the system time zone.

Reply:
OK.